### PR TITLE
feat(kuma-cp) introduce Health Discovery Service (HDS)

### PR DESCRIPTION
**Pull request by lobkovilya**
### Summary

Health Discovery Service is an Envoy protocol, which allows the receiving of information about the state of particular hosts. 

In Kuma we want to utilize this protocol for health checking of application in Universal mode. Before that, we were able to check the status only for proxy, but that's not enough especially on Universal. If an application dies but the proxy is alive we still will be sending traffic to this proxy. 

Current PR introduces the initial HDS support and provides a simple TCP health check for application ports.

#### Health Discovery Service protocol

1. Right after start Envoy sends `HealthCheckRequest` with `Node` and `Capabilites` to management server
2. Management server replies with `HealthCheckSpecifier` which tells Envoy _what_ and _how_ it should check. So `HealthCheckSpecifier` contains `cluster`, `host` and `HealthChecker`. Also `HealthCheckSpecifier` contains `Interval` which defines how often Envoy should report the status.
3. Every `Interval` Envoy sends `EndpointHealthResponse` with the health status of the hosts.

#### Health Discovery Service implementation
Management server for HDS is implemented in the same fashion as xDS server. There are Snapshot, SnapshotCache, Reconciler, and etc. `Snapshot` consists of a single resource - `HealthCheckSpecifier`, which is generated based on Dataplane resource, 1 health checker per dataplane's inbound.

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### CODE REVIEW

**Comment by jakubdyszkiewicz** [comment]
Wow, that is pretty awesome!!

What is missing:
* Auth, you are sending DP token but there is no verification
* Stats (including dashboards)

Configuration of HC is hardcoded. It should be a dynamic configuration. Either with separate policy `AppHealthChecks` or by extending `Dataplane` model.

File: pkg/hds/tracker/callbacks.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+func (t *tracker) updateDataplane(streamID xds.StreamID, port uint32, ready bool) error {
+	t.RLock()
+	defer t.RUnlock()
+	dataplaneKey, hasAssociation := t.streamsAssociation[streamID]
+	if !hasAssociation {
+		return errors.Errorf("no proxy for streamID = %d", streamID)
+	}
+
```
Comment (jakubdyszkiewicz) [inline pkg/hds/tracker/callbacks.go:191]: If I understand this code, it means that for every interval (5s) we will be fetching the DP from the database (uncached).
Is there any other place in the code that modifies health? If not, we could store this data here in some map and update only if there is a change. Look at DataplaneStatusSink and DataplaneInsight, it updates only if there is a change in the insight.

File: pkg/hds/tracker/callbacks.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	changed := false
+	for _, inbound := range dp.Spec.Networking.Inbound {
+		intf := dp.Spec.Networking.ToInboundInterface(inbound)
+		if intf.WorkloadPort != port {
+			continue
+		}
+		if inbound.Health == nil || inbound.Health.Ready != ready {
+			inbound.Health = &v1alpha1.Dataplane_Networking_Inbound_Health{
```
Comment (jakubdyszkiewicz) [inline pkg/hds/tracker/callbacks.go:204]: nit: mesh_proto

File: pkg/hds/tracker/healthcheck_generator.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+func (g *SnapshotGenerator) GenerateSnapshot(node *envoy_core.Node) (util_xds_v3.Snapshot, error) {
+	proxyID, err := xds.ParseProxyIdFromString(node.Id)
+	if err != nil {
+		return nil, err
+	}
+	dpKey := model.ResourceKey{Mesh: proxyID.Mesh, Name: proxyID.Name}
+	dp := mesh.NewDataplaneResource()
+	if err := g.readOnlyResourceManager.Get(context.Background(), dp, store.GetBy(dpKey)); err != nil {
+		return nil, err
+	}
```
Comment (jakubdyszkiewicz) [inline pkg/hds/tracker/healthcheck_generator.go:44]: Ok, so the cool thing we did when we refactored XDS sync is that we went from querying single DP every 1s to only query when hash mesh is changed. Here we are once again introducing query for single DP every 1s.

Can we change the interval of snapshot generator to like 10s? I expect DP healthchecks to be almost never changed after we start it.

File: pkg/xds/bootstrap/generator.go [hunk 4]
```diff
@@ -41,6 +41,7 @@ func NewDefaultBootstrapGenerator(
 	config *bootstrap_config.BootstrapServerConfig,
 	dpServerCertFile string,
 	dpAuthEnabled bool,
+	universal bool,
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/generator.go:44]: not a blocker, but I'd rather do pass here whether the HDS is enabled and then if on that.

(you can explicitly do HDS.enabled=false with Kubernetes deployments like I commented before or you can force HDS.enabled=false on K8S with autoconfigure)

File: pkg/xds/bootstrap/template_v2.go [hunk 5]
```diff
@@ -49,6 +49,33 @@ stats_config:
...
+{{ end }}
+{{ if .CertBytes}}
+      channelCredentials:
+        sslCredentials:
+          rootCerts:
+            inlineBytes: {{ .CertBytes }}
+{{ end }}
+      statPrefix: hds
+      targetUri: {{ .XdsHost }}:{{ .XdsPort }}
+  set_node_on_first_message_only: true
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/template_v2.go:76]: nice! we should do this on XDS and SDS too.

File: test/e2e/healthcheck_universal_test.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		Eventually(func() (string, error) {
+			output, err := cluster.GetKumactlOptions().RunKumactlAndGetOutputV(Verbose, "get", "dataplane", "dp-demo-client", "-oyaml")
+			if err != nil {
+				return "", err
+			}
+			return output, nil
+		}, "10s", "1s").Should(ContainSubstring("ready: true"))
+	})
+
+})
```
Comment (jakubdyszkiewicz) [inline test/e2e/healthcheck_universal_test.go:64]: Nice! But here we are checking if we are updating the Dataplane definition.

For me the real E2E test would be to:
* spawn web
* spawn backend-1
* start stream of requests from web to backend
* spawn backend-2
* no requests are dropped, eventually requests are loadbalanced between backend-1 and backend-2.
* when service of backend-2 is down
* then eventually backend-2 is ejected and all requests are being sent to backend-1

Not a blocker for this PR to be merged, but I'd strongly consider extending the test to check such scenario.

**Review by jakubdyszkiewicz — COMMENTED** [review]
I think there is one critical thing to consider here.

One of the points of introducing this was to be able to introduce new instance of the service but only use it when it's ready. Correct me if I'm wrong, but right now when `inbound.health` is nil, then we are treating it as healthy. When you have a dataplane with `inbound.probe` it will be treated as healthy UNTIL the first probe arrives from HDS, so we will take into account this instance too soon.

We could:
* consider inbound as unhealthy when health is nil, but probe is not nil (extra method in dataplane_helpers)
* add `health.ready=false` in DataplaneManager when you are applying inbound with a probe

**Comment by jakubdyszkiewicz** [comment]
one more thing, multizone Universal.

Not sure we are doing it right now, but I think we should exclude unhealthy from available services in Ingress

File: pkg/config/dp-server/config.go [hunk 7]
```diff
@@ -61,5 +65,79 @@ func DefaultDpServerConfig() *DpServerConfig {
...
+	if h.Timeout <= 0 {
+		return errors.New("Timeout must be greater than 0s")
+	}
+	if h.Interval <= 0 {
+		return errors.New("Interval must be greater than 0s")
+	}
+	if h.NoTrafficInterval <= 0 {
+		return errors.New("NoTrafficInterval must be greater than 0s")
+	}
+	return nil
```
Comment (nickolaev) [inline pkg/config/dp-server/config.go:142]: Please make sure these are not conflicting. For example, can the timeout be lower than the check interval?

File: api/mesh/v1alpha1/dataplane.proto [hunk 8]
```diff
@@ -76,6 +77,29 @@ message Dataplane {
...
+        // unhealthy.
+        google.protobuf.UInt32Value unhealthy_threshold = 3;
+
+        // Number of consecutive healthy checks before considering a host
+        // healthy.
+        google.protobuf.UInt32Value healthy_threshold = 4;
+
+        message Tcp {}
+
+        Tcp tcp = 5;
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply api/mesh/v1alpha1/dataplane.proto:98]: I'm not sure I have enough time to do it, so yes, let's stick to TCP only in this release

File: pkg/hds/components.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	util_xds "github.com/kumahq/kuma/pkg/util/xds"
+	util_xds_v3 "github.com/kumahq/kuma/pkg/util/xds/v3"
+)
+
+var (
+	hdsServerLog = core.Log.WithName("hds-server")
+)
+
+func RegisterHDS(rt core_runtime.Runtime, grpcSrv *grpc.Server) error {
+	snapshotCache := util_xds_v3.NewSnapshotCache(false, hasher{}, util_xds.NewLogger(hdsServerLog))
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/hds/components.go:29]: If I get your question right, DP always connects to CP using 1 connection regardless of the number of inbounds

File: pkg/hds/tracker/callbacks.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/hds/tracker/callbacks.go:191]: Users can manually update Dataplane resource and wipe out `health` section. Technically we can add logic into DataplaneManager.Update to set `health`, but this means `callbacks` won't be the only place where we modify health

**Review by nickolaev — APPROVED** [review]

File: pkg/dp-server/components.go [hunk 10]
```diff
@@ -23,7 +24,7 @@ func SetupServer(rt runtime.Runtime) error {
 	if err := bootstrap.RegisterBootstrap(rt, dpServer.httpMux); err != nil {
 		return err
 	}
-	if rt.Config().Environment == core.UniversalEnvironment && rt.Config().DpServer.Hds.Enabled {
+	if rt.Config().Mode != core.Global && rt.Config().DpServer.Hds.Enabled {
```
Comment (jakubdyszkiewicz) [inline pkg/dp-server/components.go]: nit: DP server is not run on Global at all. No reason to do the check here

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**